File: src/deadlock_build_sync/policy_cards.py
from __future__ import annotations

from dataclasses import dataclass

from .mechanics import (
    BASE_INVENTORY_SLOTS,
)
from .policy_claims import ClaimClass, PolicyError


@dataclass(frozen=True)
class CounterCard:
    threat: str
    item_id: int
    comparator_item_id: int
    mechanic_ref: str
    legal_timing: str
    alternative: str
    replacement: str
    execution_mode: str
    failure_condition: str
    evidence_ref: str
    enemy_hero_id: int | None = None
    enemy_scope: str = "whole_enemy_team"
    phase: int = 0
    tier: int = 1
    enemy_mechanics_refs: tuple[str, ...] = ()

    def __post_init__(self) -> None:
        """Require the complete mechanics-first counter contract.

        Raises:
            PolicyError: If any counter decision field is empty or invalid.

        """
        values = (
            self.threat,
            self.mechanic_ref,
            self.legal_timing,
            self.alternative,
            self.replacement,
            self.execution_mode,
            self.failure_condition,
            self.evidence_ref,
        )
        if (
            self.item_id <= 0
            or self.comparator_item_id <= 0
            or self.comparator_item_id == self.item_id
            or not all(value.strip() for value in values)
        ):
            raise PolicyError(
                "counter card is missing a mechanics-first contract field"
            )
        if self.enemy_hero_id is not None and self.enemy_hero_id <= 0:
            raise PolicyError("counter card enemy hero id must be positive")
        if (
            self.enemy_scope not in {"same_lane", "whole_enemy_team"}
            or not 0 <= self.phase <= 3
            or not 1 <= self.tier <= 4
            or not self.enemy_mechanics_refs
        ):
            raise PolicyError("counter card has invalid enemy-state evidence")

    def as_dict(self) -> dict[str, object]:
        """Return the complete serializable decision contract.

        Returns:
            A JSON-compatible counter-card object.

        """
        from .policy_codec import unstructure_counter_card

        return unstructure_counter_card(self)

    @classmethod
    def from_dict(cls, value: dict[str, object]) -> CounterCard:
        """Decode a complete counter card.

        Returns:
            The validated counter card.

        Raises:
            PolicyError: If the object is incomplete or malformed.

        """
        from .policy_codec import structure_counter_card

        return structure_counter_card(value)


@dataclass(frozen=True)
class CoreAlternativeCard:
    item_id: int
    comparator_item_id: int
    stage: int
    vs: str
    why: str
    swap: str
    when: str
    skip: str
    mechanics_refs: tuple[str, ...]
    comparator_mechanics_refs: tuple[str, ...]
    evidence_ref: str
    support: int
    effective_support: float
    overlap: float
    interval: tuple[float, float]
    fold_estimates: dict[str, float]

    def __post_init__(self) -> None:
        """Require an estimable, mechanics-grounded CORE substitution card.

        Raises:
            PolicyError: If the card fails its identity or evidence contract.

        """
        text = (
            self.vs,
            self.why,
            self.swap,
            self.when,
            self.skip,
            self.evidence_ref,
        )
        identity_invalid = (
            self.item_id <= 0
            or self.comparator_item_id <= 0
            or self.item_id == self.comparator_item_id
            or not 1 <= self.stage <= BASE_INVENTORY_SLOTS
        )
        evidence_invalid = (
            self.support < 20
            or self.effective_support < 20
            or not 0.5 <= self.overlap <= 1
        )
        interval_invalid = (
            self.interval[0] <= 0
            or self.interval[0] > self.interval[1]
            or self.interval[1] - self.interval[0] > 0.10
        )
        content_invalid = (
            not all(value.strip() for value in text)
            or not self.mechanics_refs
            or not self.comparator_mechanics_refs
        )
        folds_invalid = (
            not {"train", "validation"} <= set(self.fold_estimates)
            or abs(
                self.fold_estimates.get("train", 0.0)
                - self.fold_estimates.get("validation", 0.0)
            )
            > 0.05
        )
        if any((
            identity_invalid,
            evidence_invalid,
            interval_invalid,
            content_invalid,
            folds_invalid,
        )):
            raise PolicyError("core alternative card is incomplete or unsupported")


@dataclass(frozen=True)
class SpikeCard:
    name: str
    prerequisites: tuple[str, ...]
    acquisition_state: str
    mechanical_delta: str
    conversion_window: str
    failure_conditions: tuple[str, ...]
    counterplay: tuple[str, ...]
    evidence_class: ClaimClass
    confidence: float
    evidence_ref: str

    def __post_init__(self) -> None:
        """Require a state transition rather than an outcome-only peak.

        Raises:
            PolicyError: If transition evidence or confidence is missing.

        """
        if not 0 <= self.confidence <= 1:
            raise PolicyError("spike confidence must be between zero and one")
        if self.evidence_class == ClaimClass.DESCRIPTIVE and not self.mechanical_delta:
            raise PolicyError("an outcome-only maximum cannot define a power spike")
        if not (
            self.name.strip()
            and self.prerequisites
            and self.acquisition_state.strip()
            and self.mechanical_delta.strip()
            and self.conversion_window.strip()
            and self.failure_conditions
            and self.counterplay
            and self.evidence_ref.strip()
        ):
            raise PolicyError("spike card is missing a state-transition field")

File: crates/deadlock-guides/src/policy_cards.rs
use std::collections::BTreeMap;

use deadlock_data::{Error, Result};
use serde::{Deserialize, Serialize};

use crate::core_alternative::validate_positive_interval;
use crate::policy_claim::ClaimClass;

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct CounterCard {
    pub threat: String,
    pub item_id: u64,
    pub comparator_item_id: u64,
    pub mechanic_ref: String,
    pub legal_timing: String,
    pub alternative: String,
    pub replacement: String,
    pub execution_mode: String,
    pub failure_condition: String,
    pub evidence_ref: String,
    #[serde(default)]
    pub enemy_hero_id: Option<u64>,
    #[serde(default = "enemy_scope")]
    pub enemy_scope: String,
    #[serde(default)]
    pub phase: u8,
    #[serde(default = "first_tier")]
    pub tier: u8,
    #[serde(default)]
    pub enemy_mechanics_refs: Vec<String>,
}

fn enemy_scope() -> String {
    "whole_enemy_team".into()
}
const fn first_tier() -> u8 {
    1
}

impl CounterCard {
    /// # Errors
    /// Returns an error when the counter card lacks mechanics, item identity, enemy state, or execution details.
    pub fn validate(&self) -> Result<()> {
        let text = [
            &self.threat,
            &self.mechanic_ref,
            &self.legal_timing,
            &self.alternative,
            &self.replacement,
            &self.execution_mode,
            &self.failure_condition,
            &self.evidence_ref,
        ];
        if self.item_id == 0
            || self.comparator_item_id == 0
            || self.item_id == self.comparator_item_id
            || text.iter().any(|value| value.trim().is_empty())
        {
            return Err(Error::new(
                "Counter card lacks a required mechanics or decision field",
            ));
        }
        if self.enemy_hero_id == Some(0)
            || !["same_lane", "whole_enemy_team"].contains(&self.enemy_scope.as_str())
            || self.phase > 3
            || !(1..=4).contains(&self.tier)
            || self.enemy_mechanics_refs.is_empty()
        {
            return Err(Error::new("Counter card has invalid enemy state evidence"));
        }
        Ok(())
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct CoreAlternativeCard {
    pub vs: String,
    pub why: String,
    pub swap: String,
    pub when: String,
    pub skip: String,
    pub mechanics_refs: Vec<String>,
    pub comparator_mechanics_refs: Vec<String>,
    pub item_id: u64,
    pub comparator_item_id: u64,
    pub stage: u8,
    pub evidence_ref: String,
    pub support: u64,
    pub effective_support: f64,
    pub overlap: f64,
    pub interval: [f64; 2],
    pub fold_estimates: BTreeMap<String, f64>,
}

impl CoreAlternativeCard {
    /// # Errors
    /// Returns an error when the replacement identity, mechanics, support, interval, or temporal estimates fail admission.
    pub fn validate(&self) -> Result<()> {
        if self.item_id == 0
            || self.comparator_item_id == 0
            || self.item_id == self.comparator_item_id
            || !(1..=9).contains(&self.stage)
        {
            return Err(Error::new(
                "Core alternative card has an invalid replacement or stage",
            ));
        }
        if self.support < 20
            || !self.effective_support.is_finite()
            || self.effective_support < 20.0
            || !(0.5..=1.0).contains(&self.overlap)
        {
            return Err(Error::new(
                "Core alternative card has insufficient support or overlap",
            ));
        }
        validate_positive_interval(self.interval, None)?;
        if [
            &self.vs,
            &self.why,
            &self.swap,
            &self.when,
            &self.skip,
            &self.evidence_ref,
        ]
        .iter()
        .any(|text| text.trim().is_empty())
            || self.mechanics_refs.is_empty()
            || self.comparator_mechanics_refs.is_empty()
        {
            return Err(Error::new(
                "Core alternative card has incomplete text or mechanics references",
            ));
        }
        let train = self.fold_estimates.get("train");
        let validation = self.fold_estimates.get("validation");
        if self
            .fold_estimates
            .values()
            .any(|estimate| !estimate.is_finite())
            || train
                .zip(validation)
                .is_none_or(|(train, validation)| (train - validation).abs() > 0.05)
        {
            return Err(Error::new(
                "Core alternative card lacks consistent temporal estimates",
            ));
        }
        Ok(())
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct SpikeCard {
    pub name: String,
    pub prerequisites: Vec<String>,
    pub acquisition_state: String,
    pub mechanical_delta: String,
    pub conversion_window: String,
    pub failure_conditions: Vec<String>,
    pub counterplay: Vec<String>,
    pub evidence_class: ClaimClass,
    pub confidence: f64,
    pub evidence_ref: String,
}

impl SpikeCard {
    /// # Errors
    /// Returns an error when the card lacks a supported mechanical transition, execution window, or confidence value.
    pub fn validate(&self) -> Result<()> {
        if !(0.0..=1.0).contains(&self.confidence) {
            return Err(Error::new("Spike confidence must be between zero and one"));
        }
        let text = [
            &self.name,
            &self.acquisition_state,
            &self.mechanical_delta,
            &self.conversion_window,
            &self.evidence_ref,
        ];
        if text.iter().any(|value| value.trim().is_empty())
            || self.prerequisites.is_empty()
            || self.failure_conditions.is_empty()
            || self.counterplay.is_empty()
        {
            return Err(Error::new(
                "Spike card lacks a required state transition field",
            ));
        }
        Ok(())
    }
}
